java/src/jmri/jmrix/lenz/XNetTrafficController.java

```diff
@@ -90,67 +92,65 @@ public void forwardReply(AbstractMRListener client, AbstractMRReply m) {
+            } else if ((mask & XNetInterface.COMMINFO)
+                    == XNetInterface.COMMINFO
+                    && (((XNetReply) m).getElement(0)
+                    == XNetConstants.LI_MESSAGE_RESPONSE_HEADER)) {
+                ((XNetListener) client).message((XNetReply) m);
+            } else if ((mask & XNetInterface.CS_INFO)
+                    == XNetInterface.CS_INFO
+                    && (((XNetReply) m).getElement(0)
+                    == XNetConstants.CS_INFO
+                    || ((XNetReply) m).getElement(0)
+                    == XNetConstants.CS_SERVICE_MODE_RESPONSE
+                    || ((XNetReply) m).getElement(0)
+                    == XNetConstants.CS_REQUEST_RESPONSE
+                    || ((XNetReply) m).getElement(0)
+                    == XNetConstants.BC_EMERGENCY_STOP)) {
+                ((XNetListener) client).message((XNetReply) m);
+            } else if ((mask & XNetInterface.FEEDBACK)
+                    == XNetInterface.FEEDBACK
+                    && (((XNetReply) m).isFeedbackMessage()
+                    || ((XNetReply) m).isFeedbackBroadcastMessage())) {
+                ((XNetListener) client).message((XNetReply) m);
+            } else if ((mask & XNetInterface.THROTTLE)
+                    == XNetInterface.THROTTLE
+                    && ((XNetReply) m).isThrottleMessage()) {
+                ((XNetListener) client).message((XNetReply) m);
+            } else if ((mask & XNetInterface.CONSIST)
+                    == XNetInterface.CONSIST
+                    && ((XNetReply) m).isConsistMessage()) {
+                ((XNetListener) client).message((XNetReply) m);
+            } else if ((mask & XNetInterface.INTERFACE)
+                    == XNetInterface.INTERFACE
+                    && (((XNetReply) m).getElement(0)
+                    == XNetConstants.LI_VERSION_RESPONSE
+                    || ((XNetReply) m).getElement(0)
+                    == XNetConstants.LI101_REQUEST)) {
                 ((XNetListener) client).message((XNetReply) m);
             }
         }
     }
 
     // We use the pollMessage routines for high priority messages.
     // This means responses to time critical messages (turnout off messages).
-    LinkedBlockingQueue<XNetMessage> highPriorityQueue = null;
-    LinkedBlockingQueue<XNetListener> highPriorityListeners = null;
-
-    public void sendHighPriorityXNetMessage(XNetMessage m, XNetListener reply) {
-        try {
-            highPriorityQueue.put(m);
-            highPriorityListeners.put(reply);
-        } catch (java.lang.InterruptedException ie) {
-            log.error("Interrupted while adding High Priority Message to Queue");
-        }
+    // PENDING: these fields should be probably made private w/ accessor to force proper synchronization for reading.
+    final LinkedBlockingQueue<XNetMessage> highPriorityQueue;
+    final LinkedBlockingQueue<XNetListener> highPriorityListeners;
```

Comment:
Question: if I understand @GuardedBy correctly, then adding the annotation would require to access those queues strictly under a monitor. But, as long as the processor is single-threaded (and it is - please correct if that's wrong assumption), `pollMessage` can just omit the locks; internal state of BlockingQueue is thread-safe and since only a single thread consumes stuff, no lock is necessary. The same applies to `pollReplyHandler`.
Just writing to the two has to be atomic, which @GuardedBy syntax cannot describe (IMHO)

Reply:
You are making the assumption that the processor is single threaded.  I don’t think we can or should do that.

This is really a producer consumer queue with multiple producers.  There may only be one consumer.

Reply:
By 'processor' I meant the consumer, in your terminology; sorry for the confusion. How many threads call `pollMessage` to extract items from the queue ?

Reply:
In theory this only one thread removes items from the queue ( the single consumer ).  Multiple threads can place items in the queue, and that should all be synchronized.

Reply:
One consumer. Understood.

I suppose I could easily do some error in the implications - so if you are willing to discuss, let's verify piece by piece. Even if the state would be OK as it is, a separate decision should define whether for the sake of readability a redundant, but immediately visible, synchronization should be added. If you feel your time is being wasted at any time, please tell me; the I am benefiting from the discussion (education, excercise on abstract example) more than the project.

First the consumer side
. 
**Accessing fields**: since the fields are now `final`, then according to JLS 17.5 ("Final field semantics"), after `XNetTrafficController` constructor finishes, any thread that can observe a reference value A of `XNetTrafficController` will observe correct (initialized) values of those final fields in A. So no sycnhronization is needed to observe initialized values.

**Accesing shared state**: since each op on `LinkedBlockingQueue` is thread-safe, no extra synchronization is needed to assure internal consistency.

**Consistency of operation sequence**: According to `LinkedBlockingQueue` semantics, each single operation is atomic and leaves a consistent state visible to everyone. Once `isEmpty` becomes true (after atomic put/offer completes), it stays consistent with contents of the queue because of internal synchronization within LinkedBlockingQueue. If an item would be added (concurrently) to the FIFO queue in between `isEmpty` and `take` executed by consumer (in this order), it will not affect the enumeration order of `take`s thanks to internal synchronization of the `LinkedBlockingQueue`.
So plain check + reading ONE of the queues should not require locking - as long as the all the consumer calls run in one and single thread.

Is the above OK ?